src/transformers/azrael/main.py

```diff
@@ -0,0 +1,170 @@
+import base64
+import logging
+from typing import List
+
+from db import DB
+from transform.covalent import Covalent
+
+_LENT_FIELDS  = ["nftAddress", "tokenId", "lentAmount", "lendingId", "lendersAddress",
+        "maxRentDuration", "dailyRentPrice", "nftPrice", "isERC721", "paymentToken"]
+
+_RENTED_FIELDS = ["lendingId", "renterAddress", "rentDuration", "rentedAt"]
+
+_RETURNED_FIELDS = ["lendingId", "returnedAt"]
+
+_LENDING_STOPPED_FIELDS =  ["lendingId", "stoppedAt"]
+
+_COLLATERAL_CLAIMED_FIELDS = ["lendingId", "claimedAt"]
+
+Event = lambda name, txn_hash, fields, values: dict(zip(['event', '_id', *fields], [name, txn_hash, *values]))
+
+
+# todo: needs to inherit an interface that implements flush
+# todo: every instance should also take the address it transforms
+# todo: as a constructor argument
+class Transformer:
+    def __init__(self, address: str, events: List[str]):
+
+        self._address = address
+
+        self._transformed = []
+
+        self._db_name = "ethereum-indexer"
+        self._collection_name = f"{address}-state"
+        self._events_of_interest = events
+
+        self._flush_state = False
+
+        self._db = DB()
+
+    # todo: type that returns transformed transaction
+    def entrypoint(self, txn) -> None:
+        """_summary_
+
+        Args:
+            txn (_type_): _description_
+        """
+
+        # 1. check if there is state in the db
+        # 2. if there is state in the db, update memory with it
+        self.update_memory_state()
+
+        # routes and performs any additional logic
+        logging.info(f'Handling transaction at: {txn["block_height"]} block')
+
+        log_events = txn["log_events"]
+        # * ensures that events are supplied in the correct order
+        log_events = sorted(log_events, key=lambda x: x["log_offset"])
+
+        for event in log_events:
+
+            # * means the event was emitted by a contract that is
+            # * not of interest
+            if event["sender_address"] != self._address.lower():
+                continue
+
+            # todo: this is not good.
+            # todo: if this were to happen in an event that pertains to
+            # todo: our address, it would corrupt the state
+            if event["decoded"] is None:
+                logging.warning(f"No name for event: {event}")
+                continue
+
+            # events: Rented, Lent, CollateralClaimed, LendingStopped, Returned
+            if event["decoded"]["name"] in self._events_of_interest:
```

Comment:
let's not access event directly like this. There should be a dataclass to compartmentalize all the event data in itself.